### PR TITLE
11 angle type

### PR DESCRIPTION
**Pull request by drpjm**
# PR Details
## Description

This PR contains the implementation of a Angle type that will be used within the model types, e.g. `CtrvState`, `CtraState`, to better manage sampling algorithms. This type assumes that angles take values from [0, 2pi). The type has several arithmetic operators that will support downstream operations, such as weighted means.

## Related Issue

New angle class: #11 

## Motivation and Context

This problem solves the issue of sampling angles when near the boundary of 0 and 2pi.

## How Has This Been Tested?

The implementation was validated via pen and paper and Python scripts. Formal unit tests for the operators were created.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md)
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by adamlm — APPROVED** [review]
Looks good to me.

**Comment by adamlm** [comment]
Migrated from #15

**Comment by msmcconnell** [comment]
Note for future reference. The angle class in this PR is designed to support circular mean for use in probabilistic sampling with the UKF prediction step. 
https://en.wikipedia.org/wiki/Circular_mean

To support the 0,2pi range required by circular mean the classes stores the angle in the complex plane
https://en.wikipedia.org/wiki/Polar_coordinate_system#Complex_numbers

File: include/cooperative_perception/angle.hpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+#include <complex>
+#include <units.h>
+#include <boost/math/constants/constants.hpp>
+#include "cooperative_perception/utils.hpp"
+
+namespace cooperative_perception
+{
+/**
+ * @brief Angle type for representing and manipulating angles
```
Comment (msmcconnell) [inline include/cooperative_perception/angle.hpp:28]: I would suggest adding a comment similar to this
https://github.com/usdot-fhwa-stol/cooperative-perception-core/pull/17#issuecomment-1371458930

Which explains the internal design decision of the complex plane

File: include/cooperative_perception/angle.hpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  friend auto operator==(const Angle& lhs, const Angle& rhs) -> bool
+  {
+    return lhs.value_ == rhs.value_;
+  }
+};
+
+inline auto operator+=(Angle& lhs, const Angle& rhs) noexcept -> Angle&
+{
+  auto angle_sum = lhs.get_angle() + rhs.get_angle();
+  lhs.set_angle(angle_sum);
```
Comment (msmcconnell) [inline include/cooperative_perception/angle.hpp:65]: Not a huge issue but this is a very computationally expensive way of adding what are ultimately two numbers (multiple trig calls). I assume it's required due to the storage as a complex number but wanted to double check if there's any faster way to do it.

File: include/cooperative_perception/angle.hpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+{
+  lhs /= rhs;
+  return lhs;
+}
+
+namespace utils
+{
+auto almostEqual(const Angle& lhs, const Angle& rhs) -> bool
+{
+  return lhs.get_angle() == rhs.get_angle();
```
Comment (msmcconnell) [inline include/cooperative_perception/angle.hpp:122]: This is not almost equal as described by the function name. It's an exact equality.

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: include/cooperative_perception/angle.hpp (hunk 3 above)
Comment (adamlm) [reply include/cooperative_perception/angle.hpp:122]: The `get_angle()` function returns a `units::angle::radian_t`, and the `operator==` overload checks almost equality. It's listed as "won't fix" in the project's issues. We should put a comment above the operation explaining it. Maybe something like:

```cpp
// Despite its appearance, this operation checks for almost-equality. The get_angle() function returns a 
// units::angle::radian_t type. The operator==(...) overload checks for almost-equality. See
// https://github.com/nholthaus/units/issues/118 for a discussion.
```

Comment (msmcconnell) [reply include/cooperative_perception/angle.hpp:122]: Hmm interesting. Definitely need a comment like that to explain.

File: include/cooperative_perception/angle.hpp (hunk 2 above)
Comment (adamlm) [reply include/cooperative_perception/angle.hpp:65]: I tried alternative implementations that dealt with the `std::complex` variables directly to try and avoid the getters and setters, but the [benchmarks](https://quick-bench.com/q/XdXSr_V86D7sd12kg-xv8GExBus) did not show any improvement. Some implementations were worse.

The interface isn't going to change, so I suggest we stick with this implementation for now and revisit later if we need to.

Comment (msmcconnell) [reply include/cooperative_perception/angle.hpp:65]: Sounds good thanks for checking. That's a cool little benchmarking tool!

**Comment by drpjm** [comment]
Changes made and pushed!

**Review by msmcconnell — APPROVED** [review]
LGTM!

**Pull request merged**